File: demo/KSYLiveDemo/KSYUIUtils/KSYUIView.m
//
//  UIView+Frames.m
//  KSYGPUStreamerDemo
//
//  Created by 孙健 on 16/6/21.
//  Copyright © 2016年 ksyun. All rights reserved.
//

#import "KSYUIView.h"

@implementation KSYUIView

- (CGFloat)x{
    return  self.frame.origin.x;
}
- (void)setX:(CGFloat)x{
    CGRect frame = self.frame;
    frame.origin.x = x;
    self.frame = frame;
}

- (CGFloat)y{
    return self.frame.origin.y;
}
- (void)setY:(CGFloat)y{
    CGRect frame = self.frame;
    frame.origin.y = y;
    self.frame = frame;
}

- (CGFloat)width{
    return self.size.width;
}
- (void)setWidth:(CGFloat)width{
    CGRect frame = self.frame;
    frame.size.width = width;
    self.frame = frame;
}

- (CGFloat)height{
    return self.size.height;
}
- (void)setHeight:(CGFloat)height{
    CGRect frame = self.frame;
    frame.size.height = height;
    self.frame = frame;
}
- (CGPoint)origin{
    return self.origin;
}
- (void)setOrigin:(CGPoint)origin{
    CGRect frame = self.frame;
    frame.origin = origin;
    self.frame = frame;
}

- (CGSize)size{
    return self.frame.size;
}
- (void)setSize:(CGSize)size{
    CGRect frame = self.frame;
    frame.size = size;
    self.frame = frame;
}
#pragma mark - init
- (id) initWithParent:(KSYUIView*)pView{
    self = [self init];
    if (pView){
        self.hidden = YES;
        [pView addSubview:self];
    }
    return self;
}
- (id) init {
    self = [super init];
    self.backgroundColor = [UIColor clearColor];
    self.gap = 4;
    self.btnH = 35;
    self.winWdt = self.width;
    return self;
}

#pragma mark - UI elements layout
// 每次布局前,设置默认值
- (void) layoutUI {
    self.btnH = 30;
    self.winWdt = self.width;
    self.yPos = [[UIApplication sharedApplication] statusBarFrame].size.height;
}
- (CGFloat) getXStart {
    CGFloat xPos = _gap;
    if (_yPos > self.height){
        xPos += _winWdt;
    }
    return xPos;
}

- (void) putRow:(NSArray *) subV {
    NSInteger cnt = [subV count];
    if ( cnt < 1){
        return ;
    }
    CGFloat btnW = (_winWdt/cnt) - _gap*2;
    CGFloat xPos = [self getXStart];
    CGFloat step = _gap*2+btnW;
    CGFloat yPos = _yPos > self.height ? _yPos - self.height : _yPos;
    for (id item in subV) {
        if ([item isKindOfClass:[UIView class]]){
            UIView * v = item;
            v.frame = CGRectMake(xPos, yPos, btnW, _btnH);
        }
        xPos += step;
    }
    _yPos += (_btnH + _gap);
}
- (void) putRowFit:(NSArray *) subV {
    NSInteger cnt = [subV count];
    if ( cnt < 1){
        return ;
    }
    CGFloat xPos = [self getXStart];
    CGFloat step = 0;
    CGFloat yPos = _yPos > self.height ? _yPos - self.height : _yPos;
    for (id item in subV) {
        if ([item isKindOfClass:[UIView class]]){
            UIView * v = item;
            [v sizeToFit];
            step = v.frame.size.width;
            v.frame = CGRectMake(xPos, yPos, step, _btnH);
            xPos += step;
        }
        else {
            xPos += 5;
        }
    }
    _yPos += (_btnH + _gap);
}

- (void) putRow1:(UIView*)subV {
    CGFloat yPos = _yPos > self.height ? _yPos - self.height : _yPos;
    subV.frame = CGRectMake([self getXStart], yPos, _winWdt - _gap*2, _btnH);
    _yPos += (_btnH + _gap);
}

- (void) putRow2:(UIView*)subV0
             and:(UIView*)subV1{
    CGFloat btnW = (_winWdt/2) - _gap*2;
    CGFloat y = _yPos > self.height ? _yPos - self.height : _yPos;
    CGFloat x = [self getXStart];
    subV0.frame = CGRectMake(x, y, btnW, _btnH);
    subV1.frame = CGRectMake(x+_gap*2+btnW, y, btnW, _btnH);
    _yPos += (_btnH + _gap);
}

- (void) putRow3:(UIView*)subV0
             and:(UIView*)subV1
             and:(UIView*)subV2 {
    CGFloat btnW = (_winWdt/3) - _gap*2;
    
    CGFloat x = [self getXStart];
    CGFloat y = _yPos > self.height ? _yPos - self.height : _yPos;
    CGFloat xPos[3] = {x, x+_gap*2+btnW, x+_gap*4+btnW*2};
    if (subV0){
        subV0.frame = CGRectMake(xPos[0], y, btnW, _btnH);
    }
    if (subV1) {
        subV1.frame = CGRectMake(xPos[1], y, btnW, _btnH);
    }
    if (subV2) {
        subV2.frame = CGRectMake(xPos[2], y, btnW, _btnH);
    }
    _yPos += (_btnH + _gap);
}


//(firstV 使用内容宽度, 剩余宽度全部分配给secondV)
- (void) putNarrow:(UIView*)firstV
           andWide:(UIView*)secondV {
    CGFloat x = [self getXStart];
    CGFloat y = _yPos > self.height ? _yPos - self.height : _yPos;
    [firstV sizeToFit];
    CGRect rect = firstV.frame;
    rect.origin = CGPointMake(x, y);
    rect.size.height = _btnH;
    firstV.frame = rect;
    
    CGFloat btnW = (_winWdt) - _gap*3 - rect.size.width;
    CGFloat xPos = rect.origin.x + rect.size.width + _gap;
    secondV.frame = CGRectMake( xPos, y, btnW, _btnH);
    _yPos += (_btnH + _gap);
}
//(secondV 使用内容宽度, 剩余宽度全部分配给firstV)
- (void) putWide:(UIView*)firstV
       andNarrow:(UIView*)secondV {
    CGFloat x = [self getXStart];
    CGFloat y = _yPos > self.height ? _yPos - self.height : _yPos;
    
    [secondV sizeToFit];
    CGRect rect = secondV.frame;
    rect.size.height = _btnH;
    
    CGFloat slW = (_winWdt) - _gap*3 - rect.size.width;
    rect.origin = CGPointMake(slW+x, y);
    secondV.frame = rect;
    
    rect.origin = CGPointMake(x, y);
    rect.size.width = slW;
    firstV.frame = rect;
    _yPos += (_btnH + _gap);
}

- (void) putLable:(UIView*)lbl
          andView:(UIView*)subV{
    [self putNarrow:lbl andWide:subV];
}

- (void)  putSlider:(UIView*)sl
          andSwitch:(UIView*)sw{
    [self putWide:sl andNarrow:sw];
}

#pragma mark - new and add UI elements

- (UITextField *)addTextField: (NSString*)text{
    UITextField * textF;
    textF = [[UITextField alloc] init];
    textF.text =  text;
    textF.borderStyle = UITextBorderStyleRoundedRect;
    [self addSubview:textF];
    return textF;
}

- (UIButton *)newButton: (NSString*)title{
    UIButton * button;
    button = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [button setTitle: title forState: UIControlStateNormal];
    button.backgroundColor = [UIColor lightGrayColor];
    button.alpha = 0.9;
    button.layer.cornerRadius = 10;
    button.clipsToBounds = YES;
    [self addSubview:button];
    return button;
}

// add to default respond
- (UIButton *)addButton:(NSString*)title{
    UIButton * button = [self newButton: title];
    [button addTarget:self
               action:@selector(onBtn:)
     forControlEvents:UIControlEventTouchUpInside];
    return button;
}

- (UIButton *)addButton:(NSString*)title
                 action:(SEL)action {
    UIButton * button = [self newButton: title];
    [button addTarget:self
               action:action
     forControlEvents:UIControlEventTouchUpInside];
    return button;
}

- (UISegmentedControl *)addSegCtrlWithItems: (NSArray *) items {
    UISegmentedControl * segC;
    segC = [[UISegmentedControl alloc] initWithItems:items];
    segC.selectedSegmentIndex = 0;
    segC.layer.cornerRadius = 5;
    segC.backgroundColor = [UIColor lightGrayColor];
    [segC addTarget:self
               action:@selector(onSegCtrl:)
     forControlEvents:UIControlEventValueChanged];
    [self addSubview:segC];
    return segC;
}

- (UILabel *)addLable:(NSString*)title{
    UILabel *  lbl = [[UILabel alloc] init];
    lbl.text = title;
    [self addSubview:lbl];
    lbl.backgroundColor = [UIColor colorWithWhite:0.8 alpha:0.3]; 
    return lbl;
}

- (UISwitch *)newSwitch:(BOOL) on{
    UISwitch *sw = [[UISwitch alloc] init];
    [self addSubview:sw];
    sw.on = on;
    return sw;
}

- (UISwitch *)addSwitch:(BOOL) on{
    UISwitch *switcher = [self newSwitch:on];
    [switcher addTarget:self
                 action:@selector(onSwitch:)
       forControlEvents:UIControlEventValueChanged ];
    return switcher;
}

- (UISlider *)newSliderFrom: (float) minV
                         To: (float) maxV
                       Init: (float) iniV {
    UISlider *sl = [[UISlider alloc] init];
    [self addSubview:sl];
    sl.minimumValue = minV;
    sl.maximumValue = maxV;
    sl.value = iniV;
    sl.continuous = NO;
    return sl;
}

- (UISlider *)addSliderFrom: (float) minV
                         To: (float) maxV
                       Init: (float) iniV {
    UISlider *slider = [self newSliderFrom:minV To:maxV Init:iniV];
    [slider addTarget:self
               action:@selector(onSlider:)
     forControlEvents:UIControlEventValueChanged];
    return slider;
}

- (UISlider *)addSliderFrom: (float) minV
                         To: (float) maxV
                       Init: (float) iniV
                     action: (SEL)action {
    UISlider *slider = [self newSliderFrom:minV To:maxV Init:iniV];
    [ slider addTarget:self
                action:action
      forControlEvents:UIControlEventValueChanged ];
    return slider;
}


- (KSYNameSlider *)addSliderName: (NSString*) name
                            From: (float) minV
                              To: (float) maxV
                            Init: (float) iniV {
    KSYNameSlider *sl = [[KSYNameSlider alloc] init];
    [self addSubview:sl];
    sl.slider.minimumValue = minV;
    sl.slider.maximumValue = maxV;
    sl.slider.value = iniV;
    sl.nameL.text = name;
    sl.normalValue = (iniV -minV)/maxV;
    sl.valueL.text = [NSString stringWithFormat:@"%d", (int)iniV];
    if (iniV <2){
        sl.precision = 2;
    }
    [sl.slider addTarget:self
                  action:@selector(onSlider:)
        forControlEvents:UIControlEventValueChanged ];
    weakObj(self);
    sl.onSliderBlock = ^(id sender){
        [selfWeak onSlider:sender];
    };
    return sl;
}

#pragma mark - UI respond

//UIControlEventTouchUpInside
- (IBAction)onBtn:(id)sender {
    if (_onBtnBlock) {
        _onBtnBlock(sender);
    }
    if ([self.superview isKindOfClass:[KSYUIView class]]){
        KSYUIView * v = (KSYUIView *)self.superview;
        [v onBtn:sender];
    }
}
//UIControlEventValueChanged
- (IBAction)onSwitch:(id)sender {
    if (_onSwitchBlock){
        _onSwitchBlock(sender);
    }
    if ([self.superview isKindOfClass:[KSYUIView class]]){
        KSYUIView * v = (KSYUIView *)self.superview;
        [v onSwitch:sender];
    }
}
//UIControlEventValueChanged
- (IBAction)onSlider:(id)sender {
    if (_onSliderBlock){
        _onSliderBlock(sender);
    }
    if ([self.superview isKindOfClass:[KSYUIView class]]){
        KSYUIView * v = (KSYUIView *)self.superview;
        [v onSlider:sender];
    }
}

//UIControlEventValueChanged
- (IBAction)onSegCtrl:(id)sender {
    if (_onSegCtrlBlock){
        _onSegCtrlBlock(sender);
    }
    if ([self.superview isKindOfClass:[KSYUIView class]]){
        KSYUIView * v = (KSYUIView *)self.superview;
        [v onSegCtrl:sender];
    }
}

// 获取设备的UUID
+ (NSString *) getUuid{
    return [[[[UIDevice currentDevice] identifierForVendor] UUIDString] lowercaseString];
}
@end

File: demo/KSYLiveDemo_Swift/KSYUIUtils_Swift/KSYUIView.swift
//
//  KSYUIView.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/9.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

/**
 KSY自定义视图
 
 主要增加的功能如下:
 1. 增加更方便的视图尺寸查询属性
 2. 增加新建本SDK中常用控件的方法(按钮 滑块 开关等)
 3. 绑定事件响应的回调
 4. 增加简单的布局函数, 逐行添加任意数量的控件, 等大小,等距离放置
 
 演示SDK接口的各种视图都继承自此类
 */

class KSYUIView: UIView {

    var x: CGFloat {
        get{
            return super.frame.origin.x
        }
        set{
//            self.x = newValue
            super.frame.origin.x = newValue
        }
    }
    var y: CGFloat {
        get{
            return super.frame.origin.y
        }
        set{
//            self.y = newValue
            super.frame.origin.y = newValue
        }
    }
    var width: CGFloat {
        get{
            return super.frame.size.width
        }
        set{
//            super.width = newValue
            super.frame.size.width = newValue
        }
    }
    var height: CGFloat {
        get{
            return super.frame.size.height
        }
        set{
//            self.height = newValue
            super.frame.size.height = newValue
        }
    }
    
    var origin: CGPoint {
        get{
            return super.frame.origin
        }
        set{
            super.frame.origin = newValue
        }
    }
    
    var size: CGSize {
        get{
            return super.frame.size
        }
        set{
            super.frame.size = newValue
        }
    }
    
    var gap: CGFloat = 0        // gap between btns (default 4)
    var btnH: CGFloat = 0       // button's height (default 40)
    var winWdt: CGFloat = 0     // default self.width - gap*2
    var yPos: CGFloat = 0       // default gap*5
    // 在布局函数中, 每次使用putXXX接口增加一行控件, yPos 往下增加btnH+gap
    
    var onBtnBlock: ((AnyObject) -> Void)?
    var onSwitchBlock: ((AnyObject) -> Void)?
    var onSliderBlock: ((AnyObject) -> Void)?
    var onSegCtrlBlock: ((AnyObject) -> Void)?
    
    init() {
        super.init(frame: CGRect.zero)
        backgroundColor = UIColor.clear
        gap = 4
        btnH = 35
        winWdt = width
    }
    
    init(withParent pView: KSYUIView) {
        super.init(frame: CGRect.zero)
        backgroundColor = UIColor.clear
        gap = 4
        btnH = 35
        winWdt = width
        
        self.isHidden = true
        pView.addSubview(self)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - UI elements layout
    func layoutUI() {
        btnH = 30
        winWdt = width
        yPos = UIApplication.shared.statusBarFrame.height
    }
    
    func getXStart() -> CGFloat {
        var xPos = gap
        if yPos > height {
            xPos += winWdt
        }
        return xPos
    }
    
    // Array cannot contain NSNull
    func putRow(subV: [UIView]) {
        let cnt = subV.count
        if cnt < 1 {
            return
        }
        
        let btnW = winWdt / CGFloat(cnt) - gap * 2
        var xPos = self.getXStart()
        let step = gap * 2 + btnW
        let yPos = self.yPos > height ? self.yPos - height : self.yPos
        for item in subV {
            if item.isKind(of: UIView.self) {
                item.frame = CGRect.init(x: xPos, y: yPos, width: btnW, height: btnH)
            }
            xPos += step
        }
        self.yPos += btnH + gap
    }

    func putRow(subV2: NSArray) {
        let cnt = subV2.count
        if cnt < 1 {
            return
        }
        
        let btnW = winWdt / CGFloat(cnt) - gap * 2
        var xPos = self.getXStart()
        let step = gap * 2 + btnW
        let yPos = self.yPos > height ? self.yPos - height : self.yPos
        for item in subV2 {
            if (item as AnyObject).isKind(of: UIView.self) {
                (item as! UIView).frame = CGRect.init(x: xPos, y: yPos, width: btnW, height: btnH)
            }
            xPos += step
        }
        self.yPos += btnH + gap
    }
    
    func putRow1(subV: UIView!) {
        let yPos = self.yPos > height ? self.yPos - height : self.yPos
        subV.frame = CGRect.init(x: getXStart(), y: yPos, width: winWdt - gap * 2, height: btnH)
        self.yPos += btnH + gap
    }
    
    func putRow2(subV0: UIView!, and subV1: UIView!) {
        let btnW = winWdt / 2 - gap * 2
        let y = yPos > self.height ? yPos - height : yPos
        let x = getXStart() + gap
        subV0.frame = CGRect.init(x: x, y: y, width: btnW, height: btnH)
        subV1.frame = CGRect.init(x: x + gap * 2 + btnW, y: y, width: btnW, height: btnH)
        yPos += btnH + gap
    }
    
    func putRow3(subV0: UIView!, and subV1: UIView!, and subV2: UIView!) {
        let btnW = winWdt / 3 - gap * 2
        let x = getXStart() + gap
        let y = yPos > self.height ? yPos - height : yPos
        let xPos: [CGFloat] = [x, x + gap * 2 + btnW, x + gap * 4 + btnW * 2]
        subV0.frame = CGRect.init(x: xPos[0], y: y, width: btnW, height: btnH)
        subV1.frame = CGRect.init(x: xPos[1], y: y, width: btnW, height: btnH)
        subV2.frame = CGRect.init(x: xPos[2], y: y, width: btnW, height: btnH)
        yPos += btnH + gap
    }
    
    //(firstV 使用内容宽度, 剩余宽度全部分配给secondV)
    func putNarrow(firstV: UIView, andWide secondV: UIView) {
        let x = getXStart() + gap
        let y = yPos > self.height ? yPos - height : yPos
        firstV.sizeToFit()
        firstV.frame = CGRect.init(x: x, y: y, width: firstV.frame.width, height: btnH)
        
        let btnW = winWdt - gap * 3 - firstV.frame.width
        let xPos = firstV.frame.origin.x + firstV.frame.width + gap
        secondV.frame = CGRect.init(x: xPos, y: y, width: btnW, height: btnH)
        yPos += btnH + gap
    }
    
    //(secondV 使用内容宽度, 剩余宽度全部分配给firstV)
    func putWide(firstV: UIView, andNarrow secondV: UIView) {
        let x = getXStart() + gap
        let y = yPos > height ? yPos - height : yPos
        secondV.sizeToFit()
        
        let slW = winWdt - gap * 3 - secondV.frame.width
        
        secondV.frame = CGRect.init(x: slW + x, y: y, width: secondV.frame.width, height: btnH)
        firstV.frame = CGRect.init(x: x, y: y, width: slW, height: btnH)
        
        yPos += btnH + gap
    }
    
    func putLabel(lbl: UIView, andView subV: UIView) {
        putNarrow(firstV: lbl, andWide: subV)
    }
    
    func putSlider(sl: UIView, andSwitch sw: UIView) {
        putWide(firstV: sl, andNarrow: sw)
    }
    
    // MARK: - new and add UI elements
    func addTextField(text: String) -> UITextField {
        let textF = UITextField.init()
        textF.text = text
        textF.borderStyle = UITextBorderStyle.roundedRect
        addSubview(textF)
        
        return textF
    }
    
    func newButton(title: String) -> UIButton {
        let button = UIButton.init(type: .roundedRect)
        button.setTitle(title, for: .normal)
        button.backgroundColor = UIColor.lightGray
        button.alpha = 0.9
        button.layer.cornerRadius = 10
        button.clipsToBounds = true
        addSubview(button)
        
        return button
    }
    
    func addButton(title: String) -> UIButton {
        let button = newButton(title: title)
        button.addTarget(self, action: #selector(onBtn(sender:)), for: .touchUpInside)
        
        return button
    }
    
    // button with custom action
    func addButton(title: String, action: Selector) -> UIButton {
        let button = newButton(title: title)
        button.addTarget(self, action: action, for: .touchUpInside)
        
        return button
    }
    
    func addSegCtrlWithItems(items: Array<Any>) -> UISegmentedControl {
        let segC = UISegmentedControl.init(items: items)
        segC.selectedSegmentIndex = 0
        segC.layer.cornerRadius = 5
        segC.backgroundColor = UIColor.lightGray
        segC .addTarget(self, action: #selector(onSegCtrl(sender:)), for: .valueChanged)
        
        addSubview(segC)
        
        return segC
    }
    
    func addLabel(title: String) -> UILabel {
        let lbl = UILabel.init()
        lbl.text = title
        lbl.backgroundColor = UIColor.init(white: 0.8, alpha: 0.3)
        
        addSubview(lbl)
        
        return lbl
    }
    
    func newSwitch(on: Bool) -> UISwitch {
        let sw = UISwitch.init()
        sw.isOn = on
        
        addSubview(sw)
        
        return sw
    }
    
    func addSwitch(on: Bool) -> UISwitch {
        let sw = newSwitch(on: on)
        sw .addTarget(self, action: #selector(onSwitch(sender:)), for: .valueChanged)
        
        return sw
    }
 
    func newSlider(from minV: Float,to maxV: Float, initV: Float) -> UISlider {
        let sl = UISlider.init()
        sl.minimumValue = minV
        sl.maximumValue = maxV
        sl.value = initV
        sl.isContinuous = false

        addSubview(sl)
        
        return sl
    }
    
    func addSlider(from minV: Float,to maxV: Float, initV: Float) -> UISlider {
        let sl = newSlider(from: minV, to: maxV, initV: initV)
        sl.addTarget(self, action: #selector(onSlider(sender:)), for: .valueChanged)
        
        return sl
    }
    
    func addSlider(from minV: Float,to maxV: Float,initV: Float,action:Selector) -> UISlider {
        let sl = newSlider(from: minV, to: maxV, initV: initV)
        sl.addTarget(self, action: action, for: .valueChanged)
        
        return sl
    }

    func addSlider(name nm: String, from minV: Float, to maxV: Float, initV: Float) -> KSYNameSlider {
        
        let sl = KSYNameSlider.init()
        sl.slider.minimumValue = minV
        sl.slider.maximumValue = maxV
        sl.slider.value = initV
        sl.nameL.text = nm
        sl.normalValue = (initV - minV)/maxV
        sl.valueL.text = "\(initV )"
        addSubview(sl)
        
        if initV < 2 {
            sl.precision = 2
        }
        sl.slider.addTarget(self, action: #selector(onSlider(sender:)), for: .valueChanged)
        
        sl.onSliderBlock = { [weak self] (sender) -> Void in
            self!.onSlider(sender: sender as! UIView)
        }
        
        return sl
    }
    
    // MARK: UI respond
    func onBtn(sender: AnyObject) {
        if onBtnBlock != nil {
            onBtnBlock!(sender)
        }
        if superview!.isKind(of: KSYUIView.self) {
            (superview as! KSYUIView).onBtn(sender: sender)
        }
    }
    
    func onSwitch(sender: AnyObject) {
        if onSwitchBlock != nil {
            onSwitchBlock!(sender)
        }
        if superview!.isKind(of: KSYUIView.self) {
            (superview as! KSYUIView).onSwitch(sender: sender)
        }
    }
    
    func onSlider(sender: UIView) {
        if onSliderBlock != nil {
            onSliderBlock!(sender)
        }
        if superview!.isKind(of: KSYUIView.self) {
            (superview as! KSYUIView).onSlider(sender: sender)
        }
    }
    
    func onSegCtrl(sender: AnyObject) {
        if onSegCtrlBlock != nil {
            onSegCtrlBlock!(sender)
        }
        if superview!.isKind(of: KSYUIView.self) {
            (superview as! KSYUIView).onSegCtrl(sender: sender)
        }
    }
    
    // 获取设备的UUID
    func getUuid() -> String? {
        return UIDevice.current.identifierForVendor?.uuidString.lowercased()
    }
}
